User and Topic Kubernetes domain labels

samuel-hawker opened the pull request:
### Type of change

- Enhancement / new feature

### Description

Applies kubernetes standard labels to User operator
Applies kubernetes standard labels to Topic operator
Filter out user supplied Kubernetes domain labels from the CR applies labels - i.e app.kubernetes.io/*
Related to this issue: https://github.com/strimzi/strimzi-kafka-operator/issues/2107

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

samuel-hawker commented:
Should I update the documentation to mention that labels beginning with `app.kubernetes.io/` will be filtered out?

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:
```diff
@@ -55,29 +69,16 @@
-import java.util.Locale;
-import java.util.Map;
-import java.util.Set;
-import java.util.stream.Collectors;
```
can we avoid these changes about moving imports? I guess it's related to use a different IDE?

ppatierno commented on operator-common/src/main/java/io/strimzi/operator/common/model/Labels.java:
```diff
@@ -88,16 +90,29 @@ public static String name(HasMetadata resource) {
+            .collect(Collectors.toList());
+        if (invalidLabels.size() > 0) {
+            throw new IllegalArgumentException("Labels starting with " + STRIMZI_DOMAIN + " are not allowed in Custom Resources, such labels should be removed.");
+
```
empty line

ppatierno commented on topic-operator/src/main/java/io/strimzi/operator/topic/TopicSerialization.java:
```diff
@@ -132,10 +132,16 @@ private static int getPartitions(KafkaTopic kafkaTopic) {
+            io.strimzi.operator.common.model.Labels.KUBERNETES_NAME);
+        lbls.put(io.strimzi.operator.common.model.Labels.KUBERNETES_INSTANCE_LABEL,
+            topic.getTopicName().toString());
+        lbls.put(io.strimzi.operator.common.model.Labels.KUBERNETES_MANAGED_BY_LABEL,
```
why using the entire package name?

samuel-hawker replied:
Since there are two classes in scope with the same name `Labels` in both the `operator-common` and the `topic-operator` to call the correct one I need to use the full classpath name.
Is there another way to resolve the conflict of class names without using the full classpath?

ppatierno replied:
you are right, no unluckily there is no other mechanism as "aliases" for imports (as in C#).

samuel-hawker commented:
I do agree about the exception vs filter, I discussed this with Jakub and he suggested we filter the Labels object - I might have took that too literally.
For backwards compatibility reasons though, I wonder if users would want to be told they have to remove the labels as opposed to just having them ignored.
I can either add some logging on filter, or move the kubernetes labels check to throw an exception if passed into the functions.
@ppatierno have a preference either way?

ppatierno commented:
@samuel-hawker I have no strong opinion on that. I guess that we raised an exception for "strimzi.io" labels because they are crucial to the operational knowledge of the operator so it's better avoiding having users using this prefix. Not sure it's the same for the kubernetes prefix but at least we need a warning. Let's see what @scholzj thinks as well.

scholzj commented:
My view on this:
* `strimzi.io` labels are ours and there is no obvious usecase on having them there. I would be fine if we ignore them without an exception, but I think it is a valid issue for labels which are forbidden.
* The Kubernetes labels are valid labels and having them on the CR is not necessarily an error - they might indicate that Jaeger operator created and manages this CR as indicated in #2107. However, we do not want to take these labels over because we will replace it with our own labels. So having them on the CR is not an error and no exception makes sense. We just do not want to have them there, because they do not apply anymore to any resources created by us on the basis if the CR.

So AFAIC this is correct and expected.

scholzj commented:
The PR looks good to me. But Travis seems to be failing with:
```
[[1;31mERROR[m] [1;31mFailures: [m
[[1;31mERROR[m] [1;31m  TopicOperatorIT.testKafkaTopicWithOwnerRef:292 expected:<2> but was:<5>[m
```
whcih sounds related to this (5 different labels instead of 2?). Could you have a look at this @samuel-hawker ? Thanks

samuel-hawker commented:
Thanks for the review @scholzj 
I should have mentioned that I was looking into the Travis failures, I have changed the file and was just attempting to run the integration tests locally on my Minishift (and failing..)

Will push that fix now.

ppatierno commented:
@scholzj I agree with that but about Kubernetes labels I would write at least a warning in the log to make the user aware that the operator is skipping them.

scholzj commented:
The warnings would repeat every two minutes with something what is actually kind of an expected and valid situation. Anyone using this with Jaeger would have a log full of these despite there not being any actual issue with the labels.

ppatierno commented:
Right, this is done on every reconciliation loop. So at this point just good documentation about how kubernetes labels are handled could be useful.

samuel-hawker commented:
There was an integration test that was waiting for a specific exception, which has now changed due to the labels.
I have changed waitForEvent to now accept a string that represents a regex as opposed to a hard coded string.

samuel-hawker commented:
For some reason the Travius run for OpenJDK8 failed: https://travis-ci.org/strimzi/strimzi-kafka-operator/builds/601728167
But passed for 11, it looks unrelated to anything I have changed. Is it possible to re-run this job?

ppatierno commented:
@samuel-hawker I have just restarted the job

scholzj commented:
@strimzi-ci run tests

scholzj commented:
@strimzi-ci run tests
testcase=KafkaST#testKafkaAndZookeeperScaleUpScaleDown
